---
title: "Amplicon data notes"
author: "Lexie Christopoulos"
date: "11/20/2020"
output: github_document
---
#Intro 

We explore the processed ACIDD 16S sequence using phyloseq 

#Install phyloseq 

```{r message=FALSE, warning=FALSE}
#BiocManager::install("phyloseq")
```


```{r message=FALSE, warning=FALSE}
library(tidyverse)
library(phyloseq)
library(RColorBrewer)
```

#Import Data 

```{r message=FALSE}
count.tab <- read_rds("C:/Users/aechr/Desktop/github/144l_students/Input_Data/week 6/seqtab-nochimtaxa.rds") #table of counts for each sequence in each sample 

tax.tab <- read_rds("C:/Users/aechr/Desktop/github/144l_students/Input_Data/week 6/taxa.rds") #table that matches ASV to sequence 

sample.tab <- read_rds("C:/Users/aechr/Desktop/github/144l_students/Output_Data/week4/ACIDD_Exp_Processed_DOC_BGE.rds") %>% 
  drop_na(DNA_SampleID) %>% 
  column_to_rownames(var = "DNA_SampleID")
```

#Phyloseq Object 
 We need to create a phyloseq object that merges all three datasets. Sometimes this doesn't work because of the format of the data files. Make sure all the sample names between the sampleinfo.txt and the seqtab-nochimtaxa.txt are the same. 
 
```{r}
OTU = otu_table(count.tab, taxa_are_rows = TRUE)
TAX = tax_table(tax.tab)
SAM = sample_data(sample.tab)
ps = phyloseq(OTU,TAX,SAM)
```
 

#Filter sequences 

We will filter out chloroplasts and mitochondria, because we only intended to amplify bacterial sequences. It's good to check you don't have anything lurking in the taxonmy table. 

```{r}
sub_ps <- ps %>% 
  #subset_samples(Experiment == "ASH172") %>% #use this function if you want to only include some subset of your sample set in the subsequent analysis 
  subset_taxa(Family != "mitochondria" & Order != "Cholorplast")
```

#Sample Summary 

```{r echo=FALSE, warning=FALSE, message=FALSE, fig.height=4, fig.width=6, fig.align="center"}
#Make a data frame with a column for the read counts of each sample.
sample_sum_df <- data.frame(sum = sample_sums(sub_ps))
#Histogram of sample read counts 
ggplot(sample_sum_df, aes(x = sum))+
  geom_histogram(color = "black", fill = "#377EB8", binwidth = 1000)+
  ggtitle("Distribution of sample sequencing depth")+
  xlab("Read Counts")+
  theme(axis.title = element_blank())+
  theme_bw()


```

```{r}
#mean, max, and min of sample read counts 
summary(sample_sum_df)
```

#Beta Diversity 

Beta diversity involves calculating metrics such as distances or dissimilarities based on pairwise comparisons of samples - they do not exists for a single sample, but rather only as metrics that relate samples to each other. i.e. beta diversity = patterns in community structure between samples. 

Since differences in sampling depths between samples can influence distance/dissilimarity metrics, we first need to somehow normalize the read depth across our samples. 

##Subsample

We will rarefy (random subsample with replacement) the read depth of the samples first (scale to the smallest library size). 


Read depth is an artifact of a machine made by a company in San Diego, not anything about your samples or their biology. It is totally artifactual, and controlling for artifacts is critical in science. Subsampling randomly is the simplest way to control for this, and the question is whether this is the "best" way of controlling for it. 

A strong reason to subsample is to standardize effort. The bottom line is that in all experimental design you should not be comparing things to which you devote different effort in resolution. For instance, you don't sample one site once and week and another once a month if you want to compare the dynamics between the sites. You standardize effort. 

With that said, the bigger your differential in mean (or median) read depth (reads/sample) between pre-and-post-subsampling, the greater the "effect" on beta diversity. 

Examples:
- mean reads before = 40K, mean reads after = 1k, big effect 
-mean reads before = 40K, mean reads after = 20k, small effect 
-mean reads before = 2k, mean reads after = 1k, small effect 

We will subsample to the minimum read depth of all samples and not subsample. We'll then compare the mean reads pre-and-post-subsample and also compare beta diversity patterns. 

```{r}
ps_min <- rarefy_even_depth(sub_ps, sample.size = min(sample_sums(sub_ps)))

mean(sample_sums(sub_ps))#7686
mean(sample_sums(ps_min))#6048 this is also the same as 
min(sample_sums(sub_ps))
```

Based on the mean reads pre and post-subsampling, subsampling here shouldn't have a major effect on our beta diversity patterns 

## NMDS

One of the best exploratory analyses for amplicon data is unconstrained ordinations. Here we will look at non-metric multidimensional scaling (NMDS) ordinations of our full community samples. For NMDS plots it's important to set a seed since the starting positions of smaples in the algorithm is random. 

```{r}
set.seed(1)
#Ordinate 
nmds <- ordinate(sub_ps, method = "NMDS", distance = "bray") #stress = 0.04
```

```{r}
set.seed(1)
#Ordinate 
nmds_min <- ordinate(ps_min, method = "NMDS", distance = "bray") #stress = 0.04
```


```{r echo=FALSE, warning=FALSE, message=FALSE, fig.height=4, fig.width=6, fig.align="center"}
levels <- c("Control", "Ash Leachate", "San Diego", "Santa Barbara")

nmds.plot <- plot_ordination(sub_ps, nmds, title = "NMDS")+
  geom_point(aes(fill = days, shape = factor(Treatment, levels = levels)), alpha = 0.6, stroke = 2, size = 4)+
  scale_shape_manual(values = c(21,22,23))+
  scale_fill_gradient(low = "#0db5e6", high = "#d31f2a")+
  theme_bw()

#removing one of the plotting layers (there are points within points)
nmds.plot$layers <- nmds.plot$layers[-1]
nmds.plot +
  facet_grid(~Location)+
  guides(fill = guide_colourbar(title = "Days"), shape = guide_legend(title = "Treatment"))
```


```{r echo=FALSE, warning=FALSE, message=FALSE, fig.height=4, fig.width=6, fig.align="center"}
levels <- c("Control", "Ash Leachate", "San Diego", "Santa Barbara")

nmds_min.plot <- plot_ordination(ps_min, nmds, title = "NMDS")+
  geom_point(aes(fill = days, shape = factor(Treatment, levels = levels)), alpha = 0.6, stroke = 2, size = 4)+
  scale_shape_manual(values = c(21,22,23))+
  scale_fill_gradient(low = "#0db5e6", high = "#d31f2a")+
  theme_bw()

#removing one of the plotting layers (there are points within points)
nmds_min.plot$layers <- nmds_min.plot$layers[-1]
nmds_min.plot +
  facet_grid(~Location)+
  guides(fill = guide_colourbar(title = "Days"), shape = guide_legend(title = "Treatment"))
```


NMDS plots attempt to show ordinal distances between samples as accurately as possible in two dimmensions. It is important to report the stress of these plots, because a high stress value means that the algorithms has a hard time representing the distances between samples in 2 dimensions. The stress of this plot was good - it was 0.04 (generally anything below 0.2 is considered acceptable). 

Subsampling doesn't appear to affect the patterens we see in the beta diversity. So moving forward, we will focus on the subsampled dataset. 


#Alpha Diversity 

Estimaing alpha diversity of microbial communities is *problematic* no matter what you do. 

We are going to calculate the Chao1 index for richness and the Shannon diversity index. 

**It is important to note the the alpha diveristy values are not interpreatable as "real" numbers of anything (due to the nature of amplicon data), but they can still be useful as relative metrics of comparison. If Chao1 richness goes up, but Shannon diversity goes down, it indicates that the sample may have more ASVs but it dominated by a few of them.** 

We will use the subsampled library, which retains estimates of the species abundance of the real population while standardizing sampling effort. 

Chao1 = nonparametric estimation of minimum community richness 


```{r}
richness <- estimate_richness(ps_min, measures = c("Chao1", "Shannon")) %>% 
  rownames_to_column(., var = "DNA_ID") %>% 
  mutate_at(vars(DNA_ID), str_replace_all, pattern = "171.", "171-") %>% 
  mutate_at(vars(DNA_ID), str_replace_all, pattern = "172.", "172-")
```


Let's add the sample metadata into this dataframe. 

```{r}
alphadiv <- left_join(richness, sample.tab %>% 
                        rownames_to_column(., var = "DNA_ID"))
```

```{r echo=FALSE, fig.align="center", fig.height=10, fig.width=12, message=FALSE, warning=FALSE}

#install.packages("ggpubr")
library(ggpubr)

pivot.data <- alphadiv %>% 
  select(Treatment, Location, Bottle, Timepoint, days, Chao1, Shannon) %>% 
  pivot_longer(., cols = c(Chao1, Shannon), names_to = "measure", values_to = "est") %>% 
  left_join(., alphadiv %>% 
              select(Treatment, Location, Bottle, Timepoint, days, se.chao1)) %>% 
  mutate(se.chao1 = ifelse(measure == "Chao1", se.chao1, NA))

alpha.plot <- ggboxplot(pivot.data, x = "Timepoint", y = "est", 
                        #color = "Location",
                        #pallete = c("#0db5e6, #d31f2a),
                        xlab = expression(italic(paste(""))),
                        ylab = expression(italic("Alpha Diversity Measure")),
                        add = "dotplot",
                        width = 0.2,
                        ggtheme = theme_bw())+
  stat_compare_means(label.x = "6")+
  facet_grid(measure ~ factor(Treatment, levels = levels), scales = "free")
  
alpha.plot 


```

Boxes represent the 1.5 interquartile range, with the internal solid lines representing the median. Circles represent data points, p-values are reported the non-parametric Wilcoxon test, which tests whether the means between two groups are equal (ns: p>0.05, * : p </ 0.05, **: p </0.01)

Differences in the alpha diversity indexes among conditions were tested using pairwise Wilcoxon tests; p < 0.05 was considered the threshold significance for a difference between conditions. 

From this plot, we can see within the treatments that the richness (via Chao1 index) of our samples significantly changed, while overall diveristy (via Shannon index) did not change. This suggests that while richness decreased in both the control and ash leachate treatments, the eveness was similar between the inital and final conditions.


```{r echo=FALSE, warning=FALSE, message=FALSE, fig.height=10, fig.width=12, fig.align= "center"}
alpha.plot2 <- ggboxplot(pivot.data, x = "Treatment", y = "est", 
                        #color = "Location",
                        #pallete = c("#0db5e6, #d31f2a),
                        xlab = expression(italic(paste(""))),
                        ylab = expression(italic("Alpha Diversity Measure")),
                        add = "dotplot",
                        width = 0.2,
                        ggtheme = theme_bw())+
  stat_compare_means(label.x = "Ash Leachate")+
  facet_grid(measure ~ Timepoint, scales = "free")
  
alpha.plot2 

```


From this plot, we can see between the treatments that the richness of the control samples were higher at the inital condition that the Ash Leachate, suggesting that there may have been some quality control issues as we would expect the inital samples to all have the same richness. By timepoint 6, it looks like the richness was about the same between the contorl and the Ashe Leachate. Overall, diversity was similar between the treatments at the inital condition, but not by the end of the experiment. The Ash Leachate samples at timepoint 6 may have been less even. 

**In sum, we observe that the richness similarity decreased within the treatments overtime, but overall diveristy was lower in the Ash Leachate experiments, suggesting that those communities became less rich and less even relative to the control. i.e. relative to the control, Ash Leachate samples had less ASVs and were dominated by fewer of them.** 

#Who??

Whic taxa were important? Which taxa were contributing to the change in community composition?

**Note: Recovered 16S rRNA gene copy numbers fo not equal organism abundance**

That said, we can generate a heat map of our samples showing us how the relative abundance of different taxonomic groups change...potentially giving us a visual of which taxa are most important to the alpha and beta diversity patterns we observed. 

First, we're going to generate a custom table that will be easier to work with than a phyloseq object. 

## Generate Relative Abundances 

Our data currently shows number gene copies recovered, so we'll convert to percentages (relative abundance). 

```{r}
ps_std <- transform_sample_counts(ps_min, function(x) x/sum(x))
#extract the relative abundance table and coerce into dataframe 
ps_std.tab <- as(otu_table(ps_std), "matrix")
ps_std.df <- as.data.frame(ps_std.tab)
```


##Make Table 

```{r warning=FALSE}
#first coerce the taxa into a data frame 
tax.df <- as.data.frame(tax.tab)
#then combine the data frames 
custom.tab <- tax.df %>% 
  rownames_to_column(., var = "asv") %>% 
  left_join(., ps_std.df %>% rownames_to_column(., var = "asv")) %>%
  #create a new index that combines the class, order, family, and genus values, you can play around here!! 
  mutate(#pcofg = paste(Phylum, "_", Class, "_", Order, "_", Family, "_", Genus), 
    #pcof = paste(Phylum, "_", Class, "_", Order, "_", Family,),
    pco = paste(Phylum, "_", Class, "_", Order, "_")) %>% 
  select(-c(asv:Genus)) %>% 
  select(pco, everything()) %>% 
  group_by(pco) %>% 
  #here we are combining the relative abundances based on our grouping 
  summarise_at(vars(contains(c("ASH171", "ASH172"))), sum, na.rm = TRUE) %>% 
  ungroup()

#save the row names and then make them into the column names 
colnames <- custom.tab[,1]

#transpose the dataframe so we can merge with the sample info table 
t_custom.tab <- as.data.frame(t(custom.tab[,-1]))
colnames(t_custom.tab) <- colnames$pco

#merge 
sweet.tab <- t_custom.tab %>% 
  rownames_to_column(., var = "sample") %>% 
  left_join(., sample.tab %>% rownames_to_column(., var = "sample") %>% select(sample, Experiment, Location, Bottle, Treatment, Timepoint, days, cells)) %>% 
  select(sample, Experiment:cells, everything())

relabund <- sweet.tab %>% 
  select(-c(sample:cells)) %>% 
  #remove groups that are completely absent 
  .[ , colSums(.) > 0] %>% 
  #arrange by biggest contributors 
  .[, order(colSums(-.))] %>% 
  bind_cols(sweet.tab %>% select(sample:cells), .)
```


